Zend/zend_compile.c

```diff
@@ -5341,7 +5352,13 @@ void zend_compile_foreach(zend_ast *ast) /* {{{ */
 
 	zend_end_loop(opnum_fetch, &reset_node);
 
-	opline = zend_emit_op(NULL, ZEND_FE_FREE, &reset_node, NULL);
+	zend_emit_op(NULL, ZEND_FE_FREE, &reset_node, NULL);
+
+	if (by_ref && value_ast->kind != ZEND_AST_ARRAY && !contains_append(value_ast)) {
```

Comment:
> This RFC proposes to only perform the reference unwrapping for simple variables of the form $value (“CV”), but not for complex variables.

The RFC could also clarify that arrays **containing** simple variables of the form $value (“CV”) are not affected by the RFC, and why. (It seems feasible to fetch all variable names used as reference recursively and emit multiple unwrapping operations, but are there concerns about keeping the feature set simple?)

The `as [&$v]` also seems like a (less common) possible bug source of the same nature. Even if the keys aren't deterministic or have side effects, you don't need to evaluate the keys to know which simple variables to perform reference unwrapping on

```php
php > $values = [[1]];
php > foreach ($values as [&$v]) {}
php > $v = 123;
php > echo json_encode($values);
[[123]]
```

Reply:
Yeah, this is something I'm not sure on. Given the "only simple variables" limitations, I could see this going both ways.